ledger/src/blockstore_processor.rs

```diff
@@ -624,90 +633,125 @@ fn process_entries(
+                queue_batches_with_lock_retry(
+                    bank,
+                    starting_index,
+                    transactions,
+                    &mut batches,
+                    |batches| {
                         process_batches(
                             bank,
                             replay_tx_thread_pool,
-                            &batches,
+                            batches.drain(..),
                             transaction_status_sender,
                             replay_vote_sender,
                             batch_timing,
                             log_messages_bytes_limit,
                             prioritization_fee_cache,
-                        )?;
-                        batches.clear();
-                    }
-                }
+                        )
+                    },
+                )?;
             }
         }
     }
     process_batches(
         bank,
         replay_tx_thread_pool,
-        &batches,
+        batches.drain(..),
         transaction_status_sender,
         replay_vote_sender,
         batch_timing,
         log_messages_bytes_limit,
         prioritization_fee_cache,
     )?;
     for hash in tick_hashes {
-        bank.register_tick(hash);
+        bank.register_tick(&hash);
     }
     Ok(())
 }
 
+/// If an entry can be locked without failure, the transactions are pushed
+/// as a batch to `batches`. If the lock fails, the transactions are unlocked
+/// and the batches are processed.
+/// The locking process is retried, and if it fails again the block is marked
+/// as dead.
+/// If the lock retry succeeds, them the batch is pushed into `batches`.
+fn queue_batches_with_lock_retry(
+    bank: &Bank,
+    starting_index: usize,
+    transactions: Vec<SanitizedTransaction>,
+    batches: &mut Vec<LockedTransactionsWithIndexes<SanitizedTransaction>>,
+    mut process_batches: impl FnMut(
+        &mut Vec<LockedTransactionsWithIndexes<SanitizedTransaction>>,
+    ) -> Result<()>,
+) -> Result<()> {
+    // try to lock the accounts
+    let lock_results = bank.try_lock_accounts(&transactions);
+    let first_lock_err = first_err(&lock_results);
+    if first_lock_err.is_ok() {
+        batches.push(LockedTransactionsWithIndexes {
+            lock_results,
+            transactions,
+            starting_index,
+        });
+        return Ok(());
+    }
+
+    // We need to unlock the transactions that succeeded to lock before the
+    // retry.
+    bank.unlock_accounts(transactions.iter().zip(lock_results.iter()));
+
+    // We failed to lock, there are 2 possible reasons:
+    // 1. A batch already in `batches` holds the lock.
+    // 2. The batch is "self-conflicting" (i.e. the batch has account lock conflicts with itself)
+
+    // Use the callback to process batches, and clear them.
+    // Clearing the batches will `Drop` the batches which will unlock the accounts.
+    process_batches(batches)?;
```

Comment:
how about doing `batches.drain(..)` here for separation of concern? seems the only thing the closure is doing it just `drain(..)`-ing.

Reply:
also, strictly speaking, seems `process_batches(...)` is called while locks are held by the half-failed batches. now it's called after unlocking all. how about this?

```rust
{
    defer! {
        // We need to unlock the transactions that succeeded to lock before the
        // retry.
        bank.unlock_accounts(transactions.iter().zip(lock_results.iter()));
    }
    // Use the callback to process batches, and clear them.
    // Clearing the batches will `Drop` the batches which will unlock the accounts.
    process_batches(batches)?;
}
```

Reply:
👍 had trouble originally because couldn't have a closure accepting generic arg...but I know the type, so i'll just make closure explicitly accept Drain

Reply:
8b6fd3f0f374ae3d136e29e5a8f37d051ca8886d